File: fabric-audit-agent/core/importers/csv.js
/**
 * Minimal, dependency-free CSV parser (RFC-4180-ish).
 * Handles quoted fields, embedded commas / quotes / newlines, CRLF or LF,
 * and a leading UTF-8 BOM. Values are trimmed. Header row keys each row object.
 *
 * @param {string} text
 * @returns {{ headers: string[], rows: Record<string,string>[] }}
 */
export function parseCsv(text) {
  if (typeof text !== 'string' || text.length === 0) return { headers: [], rows: [] };
  const s = text.charCodeAt(0) === 0xfeff ? text.slice(1) : text; // strip BOM

  /** @type {string[][]} */
  const records = [];
  let record = [];
  let field = '';
  let inQuotes = false;
  let started = false; // current record has seen any char/field separator
  const n = s.length;

  const endField = () => { record.push(field); field = ''; };
  const endRecord = () => { records.push(record); record = []; started = false; };

  for (let i = 0; i < n; i++) {
    const ch = s[i];
    if (inQuotes) {
      if (ch === '"') {
        if (s[i + 1] === '"') { field += '"'; i++; }   // escaped quote
        else inQuotes = false;
      } else field += ch;
      continue;
    }
    if (ch === '"') { inQuotes = true; started = true; }
    else if (ch === ',') { endField(); started = true; }
    else if (ch === '\r') { /* ignore; LF ends the record */ }
    else if (ch === '\n') { endField(); endRecord(); }
    else { field += ch; started = true; }
  }
  if (started || field.length > 0 || record.length > 0) { endField(); endRecord(); }

  // Drop blank records (single empty cell).
  const real = records.filter(r => !(r.length === 1 && r[0].trim() === ''));
  if (real.length === 0) return { headers: [], rows: [] };

  const headers = real[0].map(h => h.trim());
  const rows = real.slice(1).map(cells => {
    /** @type {Record<string,string>} */
    const obj = {};
    headers.forEach((h, idx) => { obj[h] = (cells[idx] ?? '').trim(); });
    return obj;
  });
  return { headers, rows };
}

File: fabric-audit-agent-py/fabric_audit_agent/importers/csv.py
"""Minimal, dependency-free CSV parser. Faithful port of the Node ``core/importers/csv.js``.

Handles quoted fields, embedded commas / quotes / newlines, CRLF or LF, and a leading
UTF-8 BOM. Values are trimmed. Header row keys each row dict.
"""


def parse_csv(text):
    if not isinstance(text, str) or len(text) == 0:
        return {"headers": [], "rows": []}
    s = text[1:] if text[0] == "﻿" else text   # strip BOM

    records, record, field = [], [], ""
    in_quotes = False
    started = False
    n = len(s)
    i = 0
    while i < n:
        ch = s[i]
        if in_quotes:
            if ch == '"':
                if i + 1 < n and s[i + 1] == '"':
                    field += '"'
                    i += 1                 # escaped quote
                else:
                    in_quotes = False
            else:
                field += ch
            i += 1
            continue
        if ch == '"':
            in_quotes = True
            started = True
        elif ch == ",":
            record.append(field)
            field = ""
            started = True
        elif ch == "\r":
            pass                            # ignore; LF ends the record
        elif ch == "\n":
            record.append(field)
            field = ""
            records.append(record)
            record = []
            started = False
        else:
            field += ch
            started = True
        i += 1
    if started or len(field) > 0 or len(record) > 0:
        record.append(field)
        records.append(record)

    real = [r for r in records if not (len(r) == 1 and r[0].strip() == "")]
    if not real:
        return {"headers": [], "rows": []}

    headers = [h.strip() for h in real[0]]
    rows = []
    for cells in real[1:]:
        obj = {}
        for idx, h in enumerate(headers):
            obj[h] = (cells[idx] if idx < len(cells) else "").strip()
        rows.append(obj)
    return {"headers": headers, "rows": rows}
